src/policy/marksweepspace/native_ms/block_list.rs

```diff
@@ -0,0 +1,315 @@
+                next.clear_prev_block();
+                next.store_block_list(self);
+            } else {
+                self.first = None;
+                self.last = None;
+            }
+            head.clear_next_block();
+            head.clear_prev_block();
+            Some(head)
+        } else {
+            None
+        }
+    }
+
+    /// Push block to the front of the list
+    pub fn push(&mut self, block: Block) {
+        if self.is_empty() {
+            block.clear_next_block();
+            block.clear_prev_block();
+            self.first = Some(block);
+            self.last = Some(block);
+        } else {
+            let self_head = self.first.unwrap();
+            block.store_next_block(self_head);
+            self_head.store_prev_block(block);
+            block.clear_prev_block();
+            self.first = Some(block);
+        }
+        block.store_block_list(self);
+    }
+
+    /// Moves all the blocks of `other` into `self`, leaving `other` empty.
+    pub fn append(&mut self, other: &mut BlockList) {
+        debug_assert_eq!(self.size, other.size);
+        if !other.is_empty() {
+            debug_assert!(
+                other.first.unwrap().load_prev_block().is_none(),
+                "The other list's head has prev block: prev{} -> head{}",
+                other.first.unwrap().load_prev_block().unwrap().start(),
+                other.first.unwrap().start()
+            );
+            if self.is_empty() {
+                self.first = other.first;
+                self.last = other.last;
+            } else {
+                debug_assert!(
+                    self.first.unwrap().load_prev_block().is_none(),
+                    "Current list's head has prev block: prev{} -> head{}",
+                    self.first.unwrap().load_prev_block().unwrap().start(),
+                    self.first.unwrap().start()
+                );
+                let self_tail = self.last.unwrap();
+                let other_head = other.first.unwrap();
+                self_tail.store_next_block(other_head);
+                other_head.store_prev_block(self_tail);
+                self.last = other.last;
+            }
+            let mut cursor = other.first;
+            while let Some(block) = cursor {
+                block.store_block_list(self);
+                cursor = block.load_next_block();
+            }
+            other.reset();
+        }
+    }
+
+    /// Remove all blocks
+    fn reset(&mut self) {
+        self.first = None;
+        self.last = None;
+    }
+
+    /// Lock list
+    pub fn lock(&mut self) {
+        let mut success = false;
+        while !success {
+            success = self
+                .lock
+                .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+                .is_ok();
+        }
+    }
+
+    /// Unlock list
+    pub fn unlock(&mut self) {
+        self.lock.store(false, Ordering::SeqCst);
+    }
```

Comment:
I think this function needs more comment, and probably renamed to reflect its true purpose.  It literally took me hours to figure out what this lock is for.  I looked at where each of the methods (`push`, `pop`, `append`, `remove`) are called, but most of the time they are only mutating private data structures of a mutator/allocator, and don't need the `BlockList` to be locked.

```suggestion
    /// Lock the list for parallel sweeping.
    ///
    /// During the Release phase, multiple GC worker threads can sweep chunks and release mutators
    /// at the same time, and the same `BlockList` can be reached by traversing blocks in a chunk,
    /// and also by traversing blocks held by a mutator.  This lock is necessary to prevent
    /// multiple GC workers from mutating the same `BlockList` instance.
    ///
    /// This lock should only be used for that purpose.  There should be no race outside the
    /// Release phase.
    pub fn lock_for_sweep(&mut self) {
        let mut success = false;
        while !success {
            success = self
                .sweep_lock
                .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
                .is_ok();
        }
    }

    /// Unlock the list for parallel sweeping.
    pub fn unlock_for_sweep(&mut self) {
        self.sweep_lock.store(false, Ordering::SeqCst);
    }
```

Reply:
I added comments as suggested. But I didn't change the method name.